src/bin/ccp.rs

```diff
@@ -0,0 +1,407 @@
+        None => PathBuf::from(format!(
+            "{}/.ccp",
+            std::env::var("HOME").expect("Could not find HOME in environment variables")
+        )),
+    };
+    //println!("==> root directory is {}", root.to_string_lossy().blue().bold());
+
+    // Prepare environment. Make sure root directory exists, is accessible, has a dir called lib/
+    let res = std::fs::create_dir(root.clone());
+    match res {
+        Ok(_) => {}
+        Err(ref error) if error.kind() == ErrorKind::AlreadyExists => {}
+        Err(error) => {
+            eprintln!("error {:#?}", error);
+            std::process::exit(1);
+        }
+    };
+
+    let lib_dir = Path::new(&root).join("lib").join("src");
+    let res = std::fs::create_dir_all(lib_dir);
+    match res {
+        Ok(_) => {}
+        Err(ref error) if error.kind() == ErrorKind::AlreadyExists => {}
+        Err(error) => {
+            eprintln!("error {:#?}", error);
+            std::process::exit(1);
+        }
+    };
+
+    let alg_paths = find_algs(root.clone());
+
+    match opt.cmd {
+        Subcommand::Get { url, branch } => {
+            use regex::RegexSet;
+            let set =
+                RegexSet::new(&[r#"^[a-zA-Z0-9-._]+$"#, r#"^[a-zA-Z0-9-]+/[a-zA-Z0-9-._]+$"#])
+                    .unwrap();
+            let mut url = match set.matches(&url).into_iter().collect::<Vec<_>>().get(0) {
+                Some(0) => format!("https://github.com/ccp-project/{}", url),
+                Some(1) => format!("https://github.com/{}", url),
+                None => url,
+                Some(_) => unreachable!(),
+            };
+            if &url[url.len() - 4..] != ".git" {
+                url = format!("{}.git", url);
+            }
+
+            let branch = branch.unwrap_or(String::from("master"));
+
+            println!("==> git clone {} --branch {}", url.blue().bold(), branch);
+            let _ = Command::new("git")
+                .current_dir(root.clone())
+                .arg("clone")
+                .arg("--recurse-submodules")
+                .arg(url)
+                .arg("--branch")
+                .arg(branch)
+                .output()
+                .expect("git clone");
+
+            rebuild_library(&root, alg_paths);
+        }
+        Subcommand::List {} => {
+            for Alg {
+                crate_name,
+                crate_path,
+            } in alg_paths
+            {
+                println!("{} => {}", crate_name.bold().blue(), crate_path);
+            }
+        }
+        Subcommand::Run { alg, } => {
+            let argv = std::env::args()
+                .skip_while(|a| a != "--")
+                .collect::<Vec<String>>()[1..]
+                .join(" ");
+            println!("==> running {} {}", alg.bold().blue(), argv);
+            let lib_path = Path::new(&root)
+                .join("lib")
+                .join("target")
+                .join("debug")
+                .join("libccp.so");
```

Comment:
does this have to be done per-user btw? It seems like instead, `ccp get` should put this shared library in `/usr/local/ccp` or something, and the library itself can maybe be symlinked to some standard place like `/usr/local/lib`?

Reply:
hm yeah that sounds good, i'll change